### PR TITLE
testing: added lease mechanism in GCS to protect tests from stepping on each other's toes

### PR DESCRIPTION
**Pull request by jkowalski**
This will be only configured on Travis.

### CODE REVIEW

File: tests/testlease/testlease.go [hunk 1]
```diff
@@ -0,0 +1,235 @@
...
+	w := oh.If(cond).NewWriter(ctx)
+
+	if err := json.NewEncoder(w).Encode(ld); err != nil {
+		return false, err
+	}
+
+	if err := w.Close(); err != nil {
+		if ae, ok := err.(*googleapi.Error); ok {
+			if ae.Code == http.StatusPreconditionFailed {
+				// already exists
```
Comment (julio-lopez) [inline tests/testlease/testlease.go]: The comment may be misleading, the condition could also be `GenerationMatch`

File: tests/testlease/testlease.go [hunk 2]
```diff
@@ -0,0 +1,235 @@
...
+	ctx := context.Background()
+
+	myOwner := uuid.New().String()
+
+	if err := tryAcquireLease(ctx, t, oh, myOwner, leaseDuration); err != nil {
+		t.Fatalf("unable to acquire lease: %v", err)
+	}
+
+	defer releaseLeaseIfStillOwned(ctx, t, oh, myOwner)
+	cb()
```
Comment (julio-lopez) [inline tests/testlease/testlease.go]: Not requesting a change, just an observation.
There isn't anything attempting to guarantee that `cb` does not run past the lease duration. It seems that'd it'd be to the caller to choose a lease duration that matches the test timeout.
If `cb` took a `Context`, then maybe a deadline could be set on the context and pass it down. Nevertheless it'd be up to `cb` to honor the context cancellation.

File: repo/blob/gcs/gcs_storage_test.go [hunk 3]
```diff
@@ -23,35 +30,37 @@ func TestGCSStorage(t *testing.T) {
...
 
-	ctx := testlogging.Context(t)
-	st, err := gcs.New(ctx, &gcs.Options{
-		BucketName:                   bucket,
-		ServiceAccountCredentialJSON: credData,
-	})
+	testlease.RunWithLease(t, testLeaseID, testLeaseDuration, func() {
+		ctx := testlogging.Context(t)
+		st, err := gcs.New(ctx, &gcs.Options{
+			BucketName:                   bucket,
```
Comment (julio-lopez) [inline repo/blob/gcs/gcs_storage_test.go]: If a prefix is specified, then the tests could share the bucket and not step into each other toes. Would that work?
In fact, we'd want to test when multiple repos share the same bucket.
Same for the S3 tests.

File: repo/blob/s3/s3_storage_test.go [hunk 4]
```diff
@@ -187,8 +197,10 @@ func TestS3StorageMinioSTS(t *testing.T) {
 		t.Skip("endpoint not reachable")
 	}
 
-	createBucket(t, options)
-	testStorage(t, options)
+	testlease.RunWithLease(t, testLeaseID, testLeaseDuration, func() {
```
Comment (julio-lopez) [inline repo/blob/s3/s3_storage_test.go]: It seems that it is desirable to run this one with a lease. The observation has been that either the created user or generated token disappear under load.

**Comment by jkowalski** [comment]
abandoning this, the issue with not being able to run on PRs is not solvable this way.

**Pull request closed without merging**